# Business Problem Understanding
- <b>Predicting the energy based on input variables with least possible error.

- Installing the tensorflow 
    - <b>pip install tensorflow
- Installing the keras
    - <b>pip install --upgrade keras

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn
import tensorflow as tf
import keras

In [3]:
tf.__version__

'2.18.0'

In [4]:
keras.__version__

'3.7.0'

In [5]:
df=pd.read_excel("energy.xlsx")
df

Temperature  Vaccum  Pressure  Humidity  Energy
0           14.96   41.76   1024.07     73.17  463.26
1           25.18   62.96   1020.04     59.08  444.37
2            5.11   39.40   1012.16     92.14  488.56
3           20.86   57.32   1010.24     76.64  446.48
4           10.82   37.50   1009.23     96.62  473.90
...           ...     ...       ...       ...     ...
9563        16.65   49.69   1014.01     91.00  460.03
9564        13.19   39.18   1023.67     66.78  469.62
9565        31.32   74.33   1012.92     36.48  429.57
9566        24.48   69.45   1013.86     62.39  435.74
9567        21.60   62.52   1017.23     67.87  453.28

[9568 rows x 5 columns]

In [6]:
df.head()

Temperature  Vaccum  Pressure  Humidity  Energy
0        14.96   41.76   1024.07     73.17  463.26
1        25.18   62.96   1020.04     59.08  444.37
2         5.11   39.40   1012.16     92.14  488.56
3        20.86   57.32   1010.24     76.64  446.48
4        10.82   37.50   1009.23     96.62  473.90

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9568 entries, 0 to 9567
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Temperature  9568 non-null   float64
 1   Vaccum       9568 non-null   float64
 2   Pressure     9568 non-null   float64
 3   Humidity     9568 non-null   float64
 4   Energy       9568 non-null   float64
dtypes: float64(5)
memory usage: 373.9 KB


### Data Preprocessing

In [8]:
df.isnull().sum()

Temperature    0
Vaccum         0
Pressure       0
Humidity       0
Energy         0
dtype: int64

In [9]:
x=df.drop(columns="Energy")
y=df["Energy"]

In [10]:
#Spliting the data into train set and test set 
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

In [11]:
x_train.shape

(7654, 4)

In [12]:
x_train

Temperature  Vaccum  Pressure  Humidity
496         11.22   43.13   1017.24     80.90
294         13.67   54.30   1015.92     75.42
6796        32.84   77.95   1014.68     45.80
6785        31.91   67.83   1008.76     53.22
1203        10.37   37.50   1013.19     79.25
...           ...     ...       ...       ...
7891        16.21   50.90   1012.46     84.45
9225        13.85   44.90   1019.11     76.79
4859        16.81   38.52   1018.26     75.21
3264        12.80   41.16   1022.43     86.19
2732        32.32   67.90   1006.08     37.93

[7654 rows x 4 columns]

## Modelling

- Using ML Algorithm

In [13]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
lr.fit(x_train,y_train)

print("Coeff valuesL:",lr.coef_)
print("Intercept value:",lr.intercept_)

#Prediction
ypred_train=lr.predict(x_train)
ypred_test=lr.predict(x_test)

#Evalution
from sklearn.metrics import r2_score
print("Train pred value:",r2_score(y_train,ypred_train))
print("Test pred value: ",r2_score(y_test,ypred_test))

Coeff valuesL: [-1.97313099 -0.23649993  0.06387891 -0.15807019]
Intercept value: 452.84103716163804
Train pred value: 0.9277253998587901
Test pred value:  0.9325315554761303


In [14]:
lr.predict([[15.2,50,995,80]])

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([461.93834802])

#### Now we use Deep-Learning
- Design the model+fit on train data

#### In DL,we are having the different types of models
- 1)Perceptron Model(0 Hiden Layers)
- 2)Artificial Neural network(1 or 2 Hidden Layer)
- 3)Deep Neural Network(3 or more Hidden Layers)

### 1) Perceptron Model 

In [15]:
import tensorflow as tf
import keras

In [16]:
from keras.models import Sequential
model= Sequential()
from keras.layers import Dense
model.add(Dense(input_dim=4,units=1,activation='relu',kernel_initializer='uniform'))
model.compile(optimizer='adam',loss='mean_squared_error')

C:\Users\chani\AppData\Roaming\Python\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [17]:
model.fit(x_train,y_train,batch_size=32,epochs=50)

Epoch 1/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 121065.2344
Epoch 2/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 20197.3848
Epoch 3/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 1978.6554
Epoch 4/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 479.1642
Epoch 5/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 424.9801
Epoch 6/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 413.6815
Epoch 7/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 402.7780
Epoch 8/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 393.7520
Epoch 9/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 382.9468
Epoch 10/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 370.1860
Epoch 11/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 354.2841
Epoch 12/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 332.4267
Epoch 13/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 319.0712
Epoch 14/50
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 305.1650
Epoch 15/50
240/240 ━━━

In [18]:
ypred_train=model.predict(x_train)
ypred_test=model.predict(x_test)

240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 981us/step
60/60 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step  


In [19]:
from sklearn.metrics import r2_score
print("r2_train:",r2_score(y_train,ypred_train))
print("r2_test:",r2_score(y_test,ypred_test))

r2_train: 0.9072749525155713
r2_test: 0.9112070634790594


#### Observation :
- The R2 value is Negative.
- So for the given dataset, Perceptron model is not considered.

### 2)Artificial Neural network(1 or 2 Hidden Layer)

In [20]:
ann_model=Sequential() 
ann_model.add(Dense(input_dim=4,units=8,activation='relu',kernel_initializer='uniform'))
ann_model.add(Dense(units=1,activation='relu',kernel_initializer='uniform'))
ann_model.compile(optimizer='adam',loss='mean_squared_error')
ann_model.fit(x_train,y_train,batch_size=32,epochs=100)

Epoch 1/100


C:\Users\chani\AppData\Roaming\Python\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 151780.8906
Epoch 2/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 548.3617
Epoch 3/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 429.0721
Epoch 4/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 402.8986
Epoch 5/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 377.8631
Epoch 6/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 344.7262
Epoch 7/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 321.2542
Epoch 8/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 300.1293
Epoch 9/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 269.5273
Epoch 10/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 241.7714
Epoch 11/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 217.6699
Epoch 12/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 182.9208
Epoch 13/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 166.2525
Epoch 14/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 144.6877
Epoch 15/100
240/240 ━━━

In [21]:
ypred_train=ann_model.predict(x_train)
ypred_test=ann_model.predict(x_test)

240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
60/60 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [32]:
from sklearn.metrics import mean_squared_error,r2_score
print("Train MSE:",mean_squared_error(y_train,ypred_train))
print("Test MSE:",mean_squared_error(y_test,ypred_test))
print("Train R2:",r2_score(y_train,ypred_train))
print("Test R2:",r2_score(y_test,ypred_test))

Train MSE: 26.91741570194541
Test MSE: 25.85992700389327
Train R2: 0.9074760199926896
Test R2: 0.9115863161984913


###  3)Deep Neural Network( 3 or more Hidden Layers )

In [22]:
# For 3 Hidden Layers
ann_model=Sequential() 
ann_model.add(Dense(input_dim=4,units=8,activation='relu',kernel_initializer='uniform'))
ann_model.add(Dense(units=8,activation='relu',kernel_initializer='uniform'))
ann_model.add(Dense(units=8,activation='relu',kernel_initializer='uniform'))
ann_model.add(Dense(units=1,activation='relu',kernel_initializer='uniform'))

ann_model.compile(optimizer='adam',loss='mean_squared_error')
ann_model.fit(x_train,y_train,batch_size=32,epochs=100)

Epoch 1/100


C:\Users\chani\AppData\Roaming\Python\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


240/240 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - loss: 193370.4531
Epoch 2/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 904.2245
Epoch 3/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 429.1297
Epoch 4/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - loss: 417.5945
Epoch 5/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - loss: 393.9521
Epoch 6/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 365.2375
Epoch 7/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 346.0523
Epoch 8/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 314.0296
Epoch 9/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 285.1196
Epoch 10/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 255.7995
Epoch 11/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 233.3313
Epoch 12/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 203.9259
Epoch 13/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 177.2170
Epoch 14/100
240/240 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 153.3861
Epoch 15/100
240/240 ━━━

In [23]:
ypred_train=ann_model.predict(x_train)
ypred_test=ann_model.predict(x_test)

240/240 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
60/60 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [24]:
from sklearn.metrics import mean_squared_error,r2_score
print("Train MSE:",mean_squared_error(y_train,ypred_train))
print("Test MSE:",mean_squared_error(y_test,ypred_test))
print("Train R2:",r2_score(y_train,ypred_train))
print("Test R2:",r2_score(y_test,ypred_test))

Train MSE: 30.217144107220875
Test MSE: 28.95907330185533
Train R2: 0.8961337719708188
Test R2: 0.9009905035807146


### Observation :
- In Perceptron Model(0 Hidden Layer):
    - Train R2: -709.4346896836322
    - Test R2 : -706.6018862651368
- In ANN(using 1 Hidden Layer):
    - Train R2: 0.9074760199926896
    - Test R2: 0.9115863161984913
- In Deep Neural Network(Using More than 2 Hidden Layer):
    - Train R2: 0.8941387399182111
    - Test R2: 0.8980740770926257

#### So finally we conclude that the <b>ANN Model is the Best Model</b> for the given dataset because it is having the least error.